Pricing Page rework: Make View URL accessible

vikneshwar opened the pull request:
#### Proposed Changes

* Make View accessible from query params 
* Update the type for query params props 
* Use a flag to decide if the URL should be updated on the view switch 

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Boot up this PR 
    * Run `git fetch && git checkout add/add/url-accessible-view`
    * Run `yarn start-jetpack-cloud`
    * Goto [http://jetpack.cloud.localhost:3000/pricing?flags=jetpack/pricing-page-rework-v1](http://jetpack.cloud.localhost:3000/pricing?flags=jetpack/pricing-page-rework-v1)
* Or click on Jetpack Cloud live link below
    * Goto `/pricing?flags=jetpack/pricing-page-rework-v1`
* Confirm that the Products view is loaded by default
* Now add `&view=bundles` to the end of the URL and reload 
* Confirm that the Bundles view is loaded by default 
* Now add `&view=products` to the end of the URL and reload 
* Confirm that the Products view is loaded by default 

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202796695664057

vikneshwar committed:
Use query params to switch views in the new pricing page

- use query params to update the initial view state
- update type for the query params

jkguidaven commented:
Hi @vikneshwar, should we also consider updating the query param 'view' every time we switch the filter? I was thinking of a scenario where I go to the bundle page and I try to refresh the page. I expect to go back to the bundle page.

I dunno if this is part of the requirements, but we can maybe confirm this with the design team. wdyt?

jkguidaven commented:
Is there also a way to make the `view` field in the `ProductStoreQueryArgs` class not required? I think this is causing the unit test build to fail.

<img width="1413" alt="Screen Shot 2022-08-26 at 4 07 07 PM" src="https://user-images.githubusercontent.com/56598660/186854589-618c47fe-4493-4f9c-bc9a-7a704a8e1cfa.png">

manzoorwanijk commented on client/my-sites/plans/jetpack-plans/product-store/index.tsx:
```diff
@@ -21,7 +21,12 @@ const ProductStore: React.FC< ProductStoreProps > = ( {
 	const productSlugs = useProductSlugs( { siteId, duration } );
 
-	const [ currentView, setCurrentView ] = useState< ViewType >( 'products' );
+	const defaultViewType =
+		urlQueryArgs?.view && [ 'products', 'bundles' ].includes( urlQueryArgs.view )
+			? urlQueryArgs.view
+			: 'products';
+
+	const [ currentView, setCurrentView ] = useState< ViewType >( defaultViewType );
```
What about moving this computation inside a callback to `useState`?

```suggestion
	const [ currentView, setCurrentView ] = useState< ViewType >( () => {
		return urlQueryArgs?.view && [ 'products', 'bundles' ].includes( urlQueryArgs.view )
			? urlQueryArgs.view
			: 'products';
	} );
```

vikneshwar replied:
The results of this computation is required in other components after the recent commit, so not moving it into callback

manzoorwanijk commented:
> Hi @vikneshwar, should we also consider updating the query param 'view' every time we switch the filter? I was thinking of a scenario where I go to the bundle page and I try to refresh the page. I expect to go back to the bundle page.
> 

I agree with @jkguidaven. I would expect the URL reflect the changes when the user changes the view type.

Should we consider using URL hash instead of query params?

manzoorwanijk commented on client/my-sites/plans/jetpack-plans/product-store/types.ts:
```diff
@@ -1,23 +1,27 @@
 }
 
-export interface ProductStoreProps extends Pick< BasePageProps, 'urlQueryArgs' > {
+export interface ProductStoreQueryArgs extends QueryArgs {
+	view: ViewType;
+}
```
This should fix your TS error.

```suggestion
export type ProductStoreQueryArgs = QueryArgs & {
	view?: ViewType;
};
```

vikneshwar commented:
> I dunno if this is part of the requirements, but we can maybe confirm this with the design team. wdyt?

@jkguidaven I remember discussing during our dev meeting that it is not required, but I get your use case and why it is good to have 💯. Thanks for bringing this up, appreciate that 👍  I will make changes to update the query params

vikneshwar committed:
Pricing page update fix view filter to update url on view switch

* Update type for ProductStore
* Update Url to reflect the current view
* Pass props to view filter which decides should url be updated

vikneshwar committed:
Make pricing page main container layout as flex

elliottprogrammer commented on client/my-sites/plans/jetpack-plans/product-store/index.tsx:
```diff
@@ -32,7 +36,11 @@ const ProductStore: React.FC< ProductStoreProps > = ( {
+			<ViewFilter
+				currentView={ currentView }
+				setCurrentView={ setCurrentView }
+				shouldUpdateUrl={ isValidViewQuery }
```
I don't think this `shouldUpdateUrl` prop is needed and it can be removed because we want the `view` url query arg to always update _every_ time the user selects a view. 
Right now the way it works is the view url query only updates if the user lands on the pricing page with the view url query already present, but this is not ideal. The view url query should be added or updated _any_ time that either 'Products' or 'Bundles' is selected by the user.

elliottprogrammer commented on client/my-sites/plans/jetpack-plans/product-store/types.ts:
```diff
@@ -1,26 +1,31 @@
 export interface ViewFilterProps {
 	currentView: ViewType;
 	setCurrentView: ( currentView: ViewType ) => void;
+	shouldUpdateUrl: boolean;
```
```suggestion
```

elliottprogrammer commented on client/my-sites/plans/jetpack-plans/product-store/view-filter.tsx:
```diff
@@ -1,23 +1,32 @@
+export const ViewFilter: React.FC< ViewFilterProps > = ( {
+	currentView,
+	setCurrentView,
+	shouldUpdateUrl,
```
```suggestion
```

elliottprogrammer commented on client/my-sites/plans/jetpack-plans/product-store/view-filter.tsx:
```diff
@@ -1,23 +1,32 @@
 				<SegmentedControl.Item
 					onClick={ () => setCurrentView( 'products' ) }
 					selected={ currentView === 'products' }
+					{ ...( shouldUpdateUrl && { path: addQueryArgs( { view: 'products' }, currentPath ) } ) }
```
```suggestion
					path={ addQueryArgs( { view: 'products' }, currentPath ) }
```

elliottprogrammer commented on client/my-sites/plans/jetpack-plans/product-store/view-filter.tsx:
```diff
@@ -1,23 +1,32 @@
 				<SegmentedControl.Item
 					onClick={ () => setCurrentView( 'bundles' ) }
 					selected={ currentView === 'bundles' }
+					{ ...( shouldUpdateUrl && { path: addQueryArgs( { view: 'bundles' }, currentPath ) } ) }
```
```suggestion
					path={ addQueryArgs( { view: 'bundles' }, currentPath ) }
```

elliottprogrammer reviewed (commented):
I'm seeing an issue here..  When I initially land on `/pricing` page (no url query), and then I select "Bundles" (or "Products"), the url does not update to `/pricing?view=bundles` (or `?view=products`), which _it should_ though..
However, when I initially land on `/pricing?view=bundles` (Note, the _view_ query arg is already present), and then I select "Products" (or "Bundles") filter , it then works correctly and the url query is updated.
The `view` url query should be added or updated _**every**_ time that either 'Products' or 'Bundles' is selected by the user, regardless of whether the `view` url query is initially there or not.  What do you think? 
To fix this behavior, all we need to do is remove the `shouldUpdateUrl` prop from the ViewFilter component. I don't see why otherwise it is needed.  Comments & code suggestion below. 🙂

vikneshwar commented:
@elliottprogrammer Thanks much for the feedback, I initially made changes to update the URL on all switches no matter whether they land with the query param or not, but later while testing, I saw the use case where they land initially without params and switches to "Bundles" view, a query param did show up all of a sudden in the URL, which to me felt not a great experience for the user. So, I added logic to handle the update only if query param is already there while landing. 

If you feel that's not ideal, I can revert back that change. Thanks.

elliottprogrammer commented:
> but later while testing, I saw the use case where they land initially without params and switches to "Bundles" view, a query param did show up all of a sudden in the URL,

Thanks @vikneshwar , yes unless I'm missing something, I'm pretty sure we need to revert that change.
Picture this.. a user lands on on the `/pricing` page without any query param.  They then want to see all the bundles, so they click "Bundles". Now they want to share this page (all the bundles) with a friend or colleague of theirs. So they copy the url and paste it into Slack (or email or whatever) and send it.  The way it works right now, the link they would be sharing is `/pricing`, with no query param for `?view=bundles`, so this wouldn't work.  The colleague would click the link but they would not be directed to viewing the Bundles, they would be viewing Products.  So this is what we need to fix.

I guess I'm not understanding what's wrong with a query param showing up all of a sudden in the url, and how that is not a great experience for the user? I think it is exactly the experience we want for the user. 😉

vikneshwar commented:
@elliottprogrammer Thanks for helping me picture this case, my initial assumption while I started working on this fix is that we add support for query params so that they can land directly in a particular section from any other linked main page or in some other special case but in a normal ideal scenario most user won't be landing without query params and so my assumption was to not to show up in that situation, during this process I never had thought about user sharing pages. 

My bad I should have questioned before starting with this task on what's the exact use case for adding params. I will make sure to do that in my future task.

vikneshwar committed:
Project Pricing Page - fix the logic that updates query param on view change

Previous logic is that the query param is updated to reflect the view omly when user lands intially on the page with the query param, if they are not then dont update url on view change, the current commit updates the url not matter if user lands with the query param or not

elliottprogrammer commented:
> My bad I should have questioned before starting with this task on what's the exact use case for adding params. I will make sure to do that in my future task.

Oh no worries at all @vikneshwar! It's no problem clarifying things in the PR comments! 😄  Other than the small change that we've been discussing, everything else looks great! Thanks for tackling this! 🙌

vikneshwar commented:
@elliottprogrammer, thanks much for clarifying. I have made changes to update query params all the time.

jkguidaven reviewed (approved):
codes LGTM, tested the live link and it is working as expected. 👍
